base/file.jl

```diff
@@ -458,6 +458,32 @@ function tempdir()
     end
 end
 
+"""
+    prepare_for_deletion(path::AbstractString)
+
+Prepares the given `path` for deletion by ensuring that all directories within that
+`path` have write permissions, so that files can be removed from them.  This is
+automatically invoked by methods such as `mktempdir()` to ensure that no matter what
+weird permissions a user may have created directories with within the temporary prefix,
+it will always be deleted.
+"""
+function prepare_for_deletion(path::AbstractString)
+    # Nothing to do for non-directories
+    if !isdir(path)
+        return
+    end
+
+    try chmod(path, filemode(path) | 0o333)
+    catch; end
+    for (root, dirs, files) in walkdir(path)
+        for dir in dirs
+            dpath = joinpath(root, dir)
+            try chmod(dpath, filemode(dpath) | 0o333)
```

Comment:
Wouldn't it be much safer only to add permissions to the user? As it is written, it seems that this could sometimes be tricked into opening up write access to a system file and thus create a security vulnerability.
```suggestion
    chmod(dpath, filemode(dpath) | 0o300)
```

Reply:
Agree. I believe I made a similar comment on some version of this.

Reply:
Yeah, this code (necessarily, intentionally) has a TOCTOU vulnerability risk, so I think we need to be careful here with your earlier comments (which I didn't read until you mentioned it). With adding world writable, I believe someone can readily convince this code to chmod the entire system to +wX (by taking advantage of the enormous delay between `readdir`, `stat`/`islink`/`isdir`, and the `chmod`/`rm` post-processing operations). While that doesn't directly lead to leaking secrets, it does mean the attacker can replace any file with a malicious one that then exfiltrates the data (thereby gaining the missing read permissions).